app/src/main/java/com/twilio/voice/quickstart/VoiceActivity.java

```diff
@@ -582,17 +590,51 @@ private void requestPermissionForMicrophone() {
         }
     }
 
+    @RequiresApi(api = VERSION_CODES.M)
+    private void requestPermissionForMicrophoneAndBluetooth() {
+        if (!hasPermissions(this, Manifest.permission.RECORD_AUDIO,
+                Manifest.permission.BLUETOOTH_CONNECT)) {
+            requestPermissions(
+                    new String[]{Manifest.permission.RECORD_AUDIO,
+                            Manifest.permission.BLUETOOTH_CONNECT},
+                    PERMISSIONS_REQUEST_CODE);
+        } else {
+            registerForCallInvites();
+        }
+    }
+
+    public static boolean hasPermissions(Context context, String... permissions) {
+        if (context != null && permissions != null) {
+            for (String permission : permissions) {
+                if (ActivityCompat.checkSelfPermission(context, permission) != PERMISSION_GRANTED) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
     @Override
     public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
         /*
          * Check if microphone permissions is granted
          */
-        if (requestCode == MIC_PERMISSION_REQUEST_CODE && permissions.length > 0) {
-            if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            if (!hasPermissions(this, Manifest.permission.RECORD_AUDIO, Manifest.permission.BLUETOOTH_CONNECT)) {
+                Snackbar.make(coordinatorLayout,
+                        "Microphone &  BLUETOOTH_CONNECT permissions needed. Please allow in your application settings.",
+                        Snackbar.LENGTH_LONG).show();
```

Comment:
This means audioswitch won't be started if the user does not grant BT permissions, which seems a bit too harsh. For example, if a user does not have any BT audio devices, it would be inappropriate to force them to grant BT permissions.
Audioswitch can still be started without BT permissions, it just won't be aware of any BT devices.